tests/core/scenario/test_scenario.py

```diff
@@ -420,3 +420,292 @@ def test_get_inputs():
+        {},
+        print,
+        [data_node_1, data_node_2],
+        [data_node_4],
+        TaskId("t1"),
+    )
+    task_2 = Task("garply", {}, print, output=[data_node_4], id=TaskId("t2"))
+    task_3 = Task("waldo", {}, print, [data_node_4], None, id=TaskId("t3"))
+    task_4 = Task("fred", {}, print, [data_node_4], [data_node_5], TaskId("t4"))
+    task_5 = Task("bob", {}, print, [data_node_7], [data_node_4], None, TaskId("t5"))
+    pipeline_1 = Pipeline("plugh", {}, [task_4, task_1, task_3], PipelineId("p1"))
+    pipeline_2 = Pipeline("hulgh", {}, [task_2, task_3], PipelineId("p2"))
+    pipeline_3 = Pipeline("qulqh", {}, [task_5, task_4], PipelineId("p3"))
+    scenario = Scenario("quest", [pipeline_1, pipeline_2, pipeline_3], {}, scenario_id=ScenarioId("s1"))
+    #  p1  s1 ---
+    #            |
+    #            |---> t1 ---|      -----> t3
+    #            |           |      |
+    #      s2 ---             ---> s4 ---> t4 ---> s7
+    #  p2  t2 ---> s4 ---> t3
+    #  p3  s8 ---> t5 ---> s4 ---> t4 ---> s7
+    assert assert_two_sorted_list_of_tasks(scenario._get_sorted_tasks(), [[task_5, task_2, task_1], [task_4, task_3]])
+
+    data_node_1 = DataNode("foo", Scope.PIPELINE, "s1")
+    data_node_2 = DataNode("bar", Scope.PIPELINE, "s2")
+    data_node_4 = DataNode("qux", Scope.PIPELINE, "s4")
+    data_node_5 = DataNode("quux", Scope.PIPELINE, "s5")
+    data_node_6 = DataNode("quuz", Scope.PIPELINE, "s6")
+    data_node_7 = DataNode("corge", Scope.PIPELINE, "s7")
+    data_node_8 = DataNode("corge", Scope.PIPELINE, "s8")
+
+    task_1 = Task(
+        "grault",
+        {},
+        print,
+        [data_node_1, data_node_2],
+        [data_node_4],
+        TaskId("t1"),
+    )
+    task_2 = Task("garply", {}, print, output=[data_node_4], id=TaskId("t2"))
+    task_3 = Task("waldo", {}, print, [data_node_4], None, id=TaskId("t3"))
+    task_4 = Task("fred", {}, print, [data_node_4], [data_node_5], TaskId("t4"))
+    task_5 = Task("bob", {}, print, [data_node_8], [data_node_4], TaskId("t5"))
+    pipeline_1 = Pipeline("plugh", {}, [task_4, task_1, task_3], PipelineId("p1"))
+    pipeline_2 = Pipeline("hulgh", {}, [task_2, task_3], PipelineId("p2"))
+    pipeline_3 = Pipeline("qulqh", {}, [task_5, task_4], PipelineId("p3"))
+    pipeline_4 = Pipeline("qulqh", {}, [task_4], PipelineId("p3"))
+    scenario = Scenario("quest", [pipeline_1, pipeline_2, pipeline_3, pipeline_4], {}, scenario_id=ScenarioId("s1"))
+    #  p1  s1 ---
+    #            |
+    #            |---> t1 ---|      -----> t3
+    #            |           |      |
+    #      s2 ---             ---> s4 ---> t4 ---> s7
+    #  p2  t2 ---> s4 ---> t3
+    #  p3  s8 ---> t5 ---> s4 ---> t4 ---> s7
+    #  p4  s4 ---> t4 ---> s7
+    print(scenario._get_sorted_tasks())
+    print([[task_5, task_2, task_1, task_4], [task_3]])
+    # assert assert_two_sorted_list_of_tasks(scenario._get_sorted_tasks(), [[task_5, task_2, task_1, task_4], [task_3]])
```

Comment:
Should we uncomment the line? Do you know why it does not work?

Reply:
Not yet unfortunately :( I'm still investigating